### DISCUSSION TITLE
Difference between `useIsFetching()` vs `useQueriClient().getQueriesData().length`

### DISCUSSION THREAD

**Discussion by seiyab** [TanStack/query#5502] [Q&A]
Hi 👋 .

I wonder about difference between `useIsFetching()` and `useQueriClient().getQueriesData().length`.
```ts
  const fetching =
    useIsFetching({
      active: true,
      predicate: (query) => query.state.status !== 'success',
    });
  const loading = useQueryClient().getQueriesData({
    active: true,
    fetching: true,
    predicate: (query) => query.state.status !== 'success',
  }).length;
  console.log('fetching', fetching, 'loading', loading)
```
```
fetching 0 loading 0
fetching 0 loading 20
fetching 20 loading 20
...
```

Do someone know the reason?

I'm using 3.39.3.

I'm sorry but I can't find minimal reproduction code so I can't provide Codesandbox.

**Comment by seiyab** [comment]
Second render shows different result on every mount.
The first one, third one and later ones works as expected as far as I know.

**Comment by TkDodo** [comment]
`useIsFetching` creates a subscription - it will re-render your component every time the value changes. `useQueryClient` just reads the static query client from react context, so `useQueriClient().getQueriesData().length` will just read the current value once. If your component re-renders for a different reason, it will also re-evaluate, but it will not re-evaluate on it's own.

**Reply by seiyab** [reply]
Thanks a lot for your response.
I understand the difference about a subscription. However, I'm still confused about why they yield different values on same rendering.
I feel that I don't understand your comment especially this phrase: "but it will not re-evaluate on it's own."
Does it mean "`useQueriClient().getQueriesData().length` doesn't re-render my component on changes of the value", or "`useIsFetching()` doesn't re-evaluate the value if the re-rendering is caused by different reason"?

**Reply by TkDodo** [reply]
it means that `useQueryClient().getQueriesData().length` doesn't re-render the component on changes of the value.

**Reply by seiyab** [reply]
I understand, thank you.
So, the original problem is still somewhat unclear, right?

**Reply by TkDodo** [reply]
I don't know. Things have changed a lot since v3, and you're not showing a runnable reproduction 🤷

**Reply by seiyab** [reply]
OK.
I'll update this discussion if
- I reproduce with Codesandbox
- Or I update react-query
- Or I find a possible reason
